### PR TITLE
approve bug modify

### PR DESCRIPTION
**Pull request by xuyp1991**

### CODE REVIEW

File: contracts/eosio.system/src/onblock.cpp [hunk 1]
```diff
@@ -81,6 +81,23 @@ namespace eosio {
...
+            s.max_size = MAX_LAST_PRODUCER_SIZE;
+            s.next_index = 1;
+            s.producers.resize(MAX_LAST_PRODUCER_SIZE);
+            s.producers[0] = bpname;
+         } );
+      }
+      else {
+         _lastproducers.modify( lastproducer_info, name{0}, [&]( auto& s ) { 
+            s.producers[s.next_index] = bpname;
+            s.next_index = ((++s.next_index) % MAX_LAST_PRODUCER_SIZE);
```
Comment (fanyang1988) [inline contracts/eosio.system/src/onblock.cpp]: need a simple way

File: contracts/eosio.pledge/src/eosio.pledge.cpp [hunk 2]
```diff
@@ -215,5 +215,56 @@ namespace eosio {
...
+      });
+
+      auto report_reward = pre_allot / 2;
+      auto approve_reward = pre_allot / (2 * (rewarder.size() - 1));
+
+      auto reward_account = rewarder[0];
+      reward_assign(deduction,reward_account,report_reward);
+
+      auto isize = rewarder.size();
+      for (int i=1;i!= isize; ++i) {
```
Comment (fanyang1988) [inline contracts/eosio.pledge/src/eosio.pledge.cpp]: use for( r : rewarder ){ ... }

File: contracts/eosio.system/include/eosio.system.hpp [hunk 3]
```diff
@@ -193,6 +198,15 @@ namespace eosio {
       uint64_t primary_key() const { return bpname; }
    };
 
+   struct [[eosio::table, eosio::contract("eosio.system")]] last_producer {
+      account_name   name = bp_producer_name.value;
```
Comment (fanyang1988) [inline contracts/eosio.system/include/eosio.system.hpp]: no need this

File: contracts/eosio.system/src/bppunish.cpp [hunk 4]
```diff
@@ -40,6 +40,9 @@ namespace eosio {
       auto punish_bp = pb_tbl.find(bpname);
       check( punish_bp != pb_tbl.end() && punish_bp->effective_block_num > curr_block_num,"the bp was not Being resolved");
 
+      auto itr = std::find_if( punish_bp->approve_bp.begin(), punish_bp->approve_bp.end(), [&](const account_name& a) { return a == approver; } );
```
Comment (fanyang1988) [inline contracts/eosio.system/src/bppunish.cpp]: not in one line

File: contracts/eosio.system/src/bppunish.cpp [hunk 5]
```diff
@@ -88,22 +90,19 @@ namespace eosio {
...
-               total_reward += pledge->pledge;
-            }
-            auto proposaler_reward = total_reward / 2 ;
-            auto approver_reward = total_reward / (2 * approve_bp_num);
-            pledge::allotreward_action temp{ eosforce::pledge_account, {  {eosforce::system_account, eosforce::active_permission} } };
-            temp.send( eosforce::block_out_pledge,bpname,punish_bp->proposaler,proposaler_reward,std::string("reward proposaler") );
-            for (int i = 0; i != isize; ++i ){
+            vector<account_name> reward_account;
+            reward_account.push_back(punish_bp->proposaler);
+            for (int i = 0; i != isize; ++i ){
```
Comment (fanyang1988) [inline contracts/eosio.system/src/bppunish.cpp]: use foreach

File: contracts/eosio.system/src/bppunish.cpp [hunk 6]
```diff
@@ -88,22 +90,19 @@ namespace eosio {
...
+            vector<account_name> reward_account;
+            reward_account.push_back(punish_bp->proposaler);
+            for (int i = 0; i != isize; ++i ){ 
                if ( is_super_bp(punish_bp->approve_bp[i]) ) {
-                  temp.send( eosforce::block_out_pledge,bpname,punish_bp->approve_bp[i],approver_reward,std::string("reward approver") );
+                  reward_account.push_back(punish_bp->approve_bp[i]);
                }
             }
+            pledge::dealreward_action temp{ eosforce::pledge_account, {  {eosforce::system_account, eosforce::active_permission} } };
+            temp.send( eosforce::block_out_pledge,bpname,reward_account,std::string("deal reward") );
```
Comment (fanyang1988) [inline contracts/eosio.system/src/bppunish.cpp]: memo is useless

File: contracts/eosio.system/src/bppunish.cpp [hunk 7]
```diff
@@ -120,4 +119,60 @@ namespace eosio {
       }
       return false;
    }
+
+   void system_contract::monitorevise( const account_name& bpname ) {
+      require_auth( get_self() );
```
Comment (fanyang1988) [inline contracts/eosio.system/src/bppunish.cpp:125]: use bpname to fix one

File: contracts/eosio.system/src/onblock.cpp (hunk 1 above)
Comment (xuyp1991) [reply contracts/eosio.system/src/onblock.cpp]: modify

**Pull request merged**